### PR TITLE
Transform OpenJDK MethodHandle methods

### PR DESCRIPTION
**Pull request by liqunl**
This PR includes changes to transform OpenJDK MethodHandle methods such that the method is performant and inliner can inline the MethodHandle chain.

The main purpose of this PR is to transform the following VM INL with known object info such that inliner can match inline target with the actual call site and continue the MethodHandle chain inlining, which is critical to the performance
 - MethodHandle.invokeBasic
 - MethodHandle.linkToXXX

Notice that MethodHandle.linkToInterface is not transformed here due to current limitation of interface call representation, and it'll be addressed later

Depends on https://github.com/eclipse/omr/pull/5745

### CODE REVIEW

**Comment by liqunl** [comment]
All the changes are ready to review except the method flag `AdapterOrLambdaForm` change. This flag is used to tell if the method is what we interested in. However, the JIT has a problem inferring all the methods, and we may need #11599. Thus, make this PR WIP. 

@0xdaryl May I ask for your review?

File: runtime/compiler/optimizer/J9TransformUtil.cpp [hunk 1]
```diff
@@ -2353,3 +2354,186 @@ J9::TransformUtil::specializeInvokeExactSymbol(TR::Compilation *comp, TR::Node *
...
+   auto symRef = node->getSymbolReference();
+   auto rm = node->getSymbol()->castToMethodSymbol()->getMandatoryRecognizedMethod();
+   switch (rm)
+      {
+      case TR::java_lang_invoke_MethodHandle_linkToStatic:
+      case TR::java_lang_invoke_MethodHandle_linkToSpecial:
+      case TR::java_lang_invoke_MethodHandle_linkToVirtual:
+         break;
+      default:
+        TR_ASSERT(false, "Unsupported method %s", symRef->getSymbol()->getResolvedMethodSymbol()->getResolvedMethod()->signature(comp->trMemory()));
```
Comment (mpirvu) [inline runtime/compiler/optimizer/J9TransformUtil.cpp:2449]: Since this line should never be reached, we could use TR_ASSERT_FATAL without the fear of extra overhead

File: runtime/compiler/optimizer/J9TransformUtil.cpp [hunk 2]
```diff
@@ -2353,3 +2354,186 @@ J9::TransformUtil::specializeInvokeExactSymbol(TR::Compilation *comp, TR::Node *
...
+   if (mhIndex == TR::KnownObjectTable::UNKNOWN ||
+       knot->isNull(mhIndex))
+      {
+      if (trace)
+         traceMsg(comp, "MethodHandle for invokeBasic n%dn %p is unknown or null\n", node->getGlobalIndex(), node);
+      return false;
+      }
+
+   TR_J9VMBase* fej9 = comp->fej9();
+   auto targetMethod = fej9->targetMethodFromMethodHandle(comp, mhIndex);
```
Comment (mpirvu) [inline runtime/compiler/optimizer/J9TransformUtil.cpp:2372]: In `targetMethodFromMethodHandle`, from KOT we get the object at the given index and from the object we get a j9method. Can the the mhIndex <--> j9method association ever change? If not, I am thinking of caching this association to avoid extra messages for JITServer. FYI @dmitry-ten

Comment (liqunl) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:2372]: That's a good question, it seems to me that the j9method of a MemberName can change due to HCR #11582. @babsingh Can you confirm?

Comment (babsingh) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:2372]: Yes, it can change due to HCR. HCR support for OJDK MHs is being tracked through https://github.com/eclipse/openj9/issues/11528.

Comment (mpirvu) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:2372]: Since KOT only exists during a compilation we should be fine because class redefinition aborts compilations in progress.

**Comment by 0xdaryl** [comment]
This is actually a breaking OMR dependency.  Updating the label.

File: runtime/compiler/optimizer/J9TransformUtil.cpp (hunk 2 above)
Comment (liqunl) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:2372]: Yes, HCR will be addressed in #11528, and the ultimate goal is to update MemberName fields in redefinition event so that we don't get stale j9method pointer

**Comment by liqunl** [comment]
@0xdaryl I update the PR to check the class flag for generated MH methods. Could you review? I'll fix the copyright issue later

**Comment by 0xdaryl** [comment]
@mpirvu : if you are satisfied with the changes can you formally approve please?

File: runtime/compiler/env/j9method.cpp [hunk 3]
```diff
@@ -204,9 +204,16 @@ TR_J9VMBase::createResolvedMethod(TR_Memory * trMemory, TR_OpaqueMethodBlock * a
...
+TR_J9VMBase::createResolvedMethodWithVTableSlot(TR_Memory * trMemory, uint32_t vTableSlot, TR_OpaqueMethodBlock * aMethod,
+                                  TR_ResolvedMethod * owningMethod, TR_OpaqueClassBlock *classForNewInstance)
+   {
+   return createResolvedMethodWithSignature(trMemory, aMethod, classForNewInstance, NULL, -1, owningMethod, vTableSlot);
+   }
+
 TR_ResolvedMethod *
 TR_J9VMBase::createResolvedMethodWithSignature(TR_Memory * trMemory, TR_OpaqueMethodBlock * aMethod, TR_OpaqueClassBlock *classForNewInstance,
-                          char *signature, int32_t signatureLength, TR_ResolvedMethod * owningMethod)
+                          char *signature, int32_t signatureLength, TR_ResolvedMethod * owningMethod, uint32_t vTableSlot)
```
Comment (mpirvu) [inline runtime/compiler/env/j9method.cpp:216]: Since the implementation of this frontend method changed, I think we need to update the corresponding implementation in JITServer

File: runtime/compiler/optimizer/MethodHandleTransformer.cpp [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+           auto mhIndex = getObjectInfoOfNode(node->getFirstArgument());
+           if (isKnownObject(mhIndex) && !knot->isNull(mhIndex))
+              {
+              TR::VMAccessCriticalSection dereferenceKnownObjectField(comp()->fej9());
+              uintptr_t mhObject = knot->getPointer(mhIndex);
+              uintptr_t mnObject = comp()->fej9()->getReferenceField(mhObject, "member", "Ljava/lang/invoke/MemberName;");
+              auto mnIndex = knot->getOrCreateIndex(mnObject);
+              if (trace())
+                 traceMsg(comp(), "Get DirectMethodHandle.member known object %d\n", mnIndex);
+              return mnIndex;
```
Comment (mpirvu) [inline runtime/compiler/optimizer/MethodHandleTransformer.cpp:335]: @dmitry-ten This sequence and the one below will need a message to the JITServer client

File: runtime/compiler/optimizer/MethodHandleTransformer.cpp [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
+          baseNode->hasUnresolvedSymbolReference())
+         return;
+
+      auto baseSymRef = baseNode->getSymbolReference();
+      LocalObjectInfo baseObj = getObjectInfoOfNode(baseNode);
+      if (trace())
+         traceMsg(comp(), "base object for indirect load n%dn is obj%d\n", node->getGlobalIndex(), baseObj);
+
+      auto knot = comp()->getKnownObjectTable();
+      if (isKnownObject(baseObj)  && !knot->isNull(baseObj))
```
Comment (mpirvu) [inline runtime/compiler/optimizer/MethodHandleTransformer.cpp:412]: We need a test on `knot != NULL` because knot can be disabled through an option, though it could be argues that `isKnownObject(baseObj)` cannot return `true` if knot doesn't exist.

File: runtime/compiler/optimizer/MethodHandleTransformer.cpp [hunk 6]
```diff
@@ -0,0 +1,562 @@
...
+         break;
+      }
+   }
+
+// Visit a node may change the object info
+//
+void
+TR_MethodHandleTransformer::visitNode(TR::TreeTop* tt, TR::Node* node, TR::NodeChecklist &visitedNodes)
+   {
+   if (visitedNodes.contains(node)) return;
```
Comment (mpirvu) [inline runtime/compiler/optimizer/MethodHandleTransformer.cpp:454]: Better to place `return` on the next line to follow our coding conventions.

File: runtime/compiler/optimizer/MethodHandleTransformer.cpp (hunk 4 above)
Comment (dmitry-ten) [reply runtime/compiler/optimizer/MethodHandleTransformer.cpp:335]: Yeah, I will change this to support JITServer in a separate PR. 
As I understand, OpenJDK method handles are not yet enabled by default, so we can leave it like this for now.

File: runtime/compiler/env/j9method.cpp (hunk 3 above)
Comment (liqunl) [reply runtime/compiler/env/j9method.cpp:220]: Fixed

**Comment by liqunl** [comment]
@mpirvu I've made changes according to your comments, could you review again?
@0xdaryl This PR is ready for your review.

File: runtime/compiler/env/VMJ9.cpp [hunk 7]
```diff
@@ -3112,6 +3112,18 @@ TR_J9VMBase::isMethodTracingEnabled(TR_OpaqueMethodBlock *method)
    return VM_VMHelpers::methodBeingTraced(_jitConfig->javaVM, (J9Method *)method);
    }
 
+bool
+TR_J9VMBase::isLambdaFormGeneratedMethod(TR_OpaqueMethodBlock *method)
+   {
+   return VM_VMHelpers::isLambdaFormGeneratedMethod(vmThread(), (J9Method *)method);
+   }
+
```
Comment (mpirvu) [inline runtime/compiler/env/VMJ9.cpp:3120]: I don't remember seeing this last time. This will require an implementation for JITServer. @dmitry-ten

Comment (dmitry-ten) [reply runtime/compiler/env/VMJ9.cpp:3120]: I don't see this method used anywhere in this PR.
Where is it supposed to be called from?

File: runtime/compiler/optimizer/J9TransformUtil.cpp (hunk 1 above)
Comment (mpirvu) [reply runtime/compiler/optimizer/J9TransformUtil.cpp:2447]: A reminder to change this into an ASSERT_FATAL

**Review by mpirvu — COMMENTED** [review]
I am ok with the implementation, though I am not the best reviewer to vouch for the correctness of the transformations in the optimizer.
We'll solve the issues  JITServer in a subsequent PR.

**Comment by mpirvu** [comment]
Linter is showing some build errors:
```
../compiler/optimizer/J9OptimizationManager.cpp:94:17: error: no member named 'methodHandleTransformer' in namespace 'OMR'; did you mean 'varHandleTransformer'?
      case OMR::methodHandleTransformer:
           ~~~~~^~~~~~~~~~~~~~~~~~~~~~~
                varHandleTransformer
../omr/compiler/optimizer/OMROptimizations.enum:84:17: note: 'varHandleTransformer' declared here
```

**Comment by dmitry-ten** [comment]
That's due to dependency on https://github.com/eclipse/omr/pull/5745

File: runtime/compiler/optimizer/J9TransformUtil.cpp (hunk 1 above)
Comment (liqunl) [reply runtime/compiler/optimizer/J9TransformUtil.cpp]: Fixed.

File: runtime/compiler/env/VMJ9.cpp (hunk 7 above)
Comment (liqunl) [reply runtime/compiler/env/VMJ9.cpp:3120]: It's called from MethodHandleTransformer.cpp

Comment (dmitry-ten) [reply runtime/compiler/env/VMJ9.cpp:3120]: Ok, I see. Yeah, in that case I'll have to override it for JITServer as well.

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk16 depends eclipse/omr#5745

**Comment by 0xdaryl** [comment]
@liqunl : please see the CI failures.  At the very least it looks like you need to `#include "il/AutomaticSymbol.hpp"` in the new MethodHandleTransformer.cpp

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk16 depends eclipse/omr#5745

**Comment by 0xdaryl** [comment]
The AIX CI build is failing.  I think you need to add a `#include "env/KnownObjectTable.hpp` to optimizer/J9TransformUtil.hpp.

**Comment by 0xdaryl** [comment]
^^^  And add it to `optimizer/MethodHandleTransformer.cpp` as well...

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk16 depends eclipse/omr#5745

**Comment by liqunl** [comment]
> The AIX CI build is failing. I think you need to add a #include "env/KnownObjectTable.hpp to optimizer/J9TransformUtil.hpp.

It's strange that "env/KnownObjectTable.hpp" is included in OMRTransformUtil.hpp, which is included in J9TransformUtil.hpp. And we've been using `TR::KnownObjectTable::UNKNOWN` before this change.

> ^^^ And add it to optimizer/MethodHandleTransformer.cpp as well...

I have "env/KnownObjectTable.hpp" included in MethodHandleTransformer.hpp which is included in MethodHandleTransformer.cpp

**Comment by liqunl** [comment]
Testing the change with inliner changes, there will be some modifications, thus WIP the PR for now.

**Comment by liqunl** [comment]
@0xdaryl Updated the commits with workaround to the xlc link error, as well as a workaround to incomplete auto symref list issue. I've also opened two issues for the two problem I'm seeing. Could you review again?

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk16 depends eclipse/omr#5745

**Comment by liqunl** [comment]
Still getting the link error on aix, it's weird that my personal build on aix passed, but the PR build still failed. Will look into this problem.

**Comment by 0xdaryl** [comment]
That is strange.  Perhaps it is the xlC version?  This is what is used in the PR build:
```
[2021-03-09T20:56:26.851Z] configure: Using xlc C compiler version 16.1.0 [IBM XL C/C++ for AIX, V16.1.0 (5725-C72, 5765-J12) Version: 16.01.0000.0000]
```

**Comment by liqunl** [comment]
@0xdaryl Static casting `TR::KnownObjectTable::UNKNOWN` fixes the xlc link error, could you review again?

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk16 depends eclipse/omr#5745

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk8,jdk11 depends eclipse/omr#5745

**Comment by 0xdaryl** [comment]
Jenkins test sanity win32 jdk8 depends eclipse/omr#5745

**Review by 0xdaryl — APPROVED** [review]
